src/freertos_drivers/common/BitBangI2C.cxx

```diff
@@ -0,0 +1,353 @@
+                        state_ = State::STOP;
+                        stateStop_ = StateStop::FIRST;
+                    }
+                    else
+                    {
+                        exit = true;
+                    }
+                }
+            }
+            break;
+        case State::STOP:
+            exit = state_stop();
+            break;
+    }
+
+    if (exit)
+    {
+        disableTick_();
+        int woken = 0;
+        sem_.post_from_isr(&woken);
+        os_isr_exit_yield_test(woken);
+    }
+}
+
+//
+// BitBangI2C::state_start()
+//
+__attribute__((optimize("-O3")))
+bool BitBangI2C::state_start()
+{
+    switch (stateStart_)
+    {
+        // start sequence
+        case StateStart::SDA_SET:
+            gpio_set(sda_);
+            ++stateStart_;
+            break;
+        case StateStart::SCL_SET:
+            gpio_set(scl_);
+            ++stateStart_;
+            break;
+        case StateStart::SDA_CLR:
+            gpio_clr(sda_);
+            ++stateStart_;
+            break;
+        case StateStart::SCL_CLR:
+            gpio_clr(scl_);
+            return true;
+    }
+    return false;
+}
+
+//
+// BitBangI2C::state_stop()
+//
+__attribute__((optimize("-O3")))
+bool BitBangI2C::state_stop()
+{
+    switch (stateStop_)
+    {
+        case StateStop::SDA_CLR:
+            gpio_clr(sda_);
+            ++stateStop_;
+            break;
+        case StateStop::SCL_SET:
+            gpio_set(scl_);
+            ++stateStop_;
+            break;
+        case StateStop::SDA_SET:
+            gpio_set(sda_);
+            stop_ = false;
+            return true; // exit
+    }
+    return false;
+}
+
+//
+// BitBangI2C::state_tx()
+//
+__attribute__((optimize("-O3")))
+bool BitBangI2C::state_tx(uint8_t data)
+{
+    switch(stateTx_)
+    {
+        case StateTx::DATA_7_SCL_SET:
+        case StateTx::DATA_6_SCL_SET:
+        case StateTx::DATA_5_SCL_SET:
+        case StateTx::DATA_4_SCL_SET:
+        case StateTx::DATA_3_SCL_SET:
+        case StateTx::DATA_2_SCL_SET:
+        case StateTx::DATA_1_SCL_SET:
+        case StateTx::DATA_0_SCL_SET:
+        {
+            uint8_t mask = 0x80 >>
+                ((static_cast<int>(stateTx_) -
+                  static_cast<int>(StateTx::DATA_7_SCL_SET)) >> 1);
+            if (data & mask)
+            {
+                gpio_set(sda_);
+            }
+            else
+            {
+                gpio_clr(sda_);
+            }
+            gpio_set(scl_);
+            ++stateTx_;
+            break;
+        }
+        case StateTx::DATA_7_SCL_CLR:
+        case StateTx::DATA_6_SCL_CLR:
+        case StateTx::DATA_5_SCL_CLR:
+        case StateTx::DATA_4_SCL_CLR:
+        case StateTx::DATA_3_SCL_CLR:
+        case StateTx::DATA_2_SCL_CLR:
+        case StateTx::DATA_1_SCL_CLR:
+        case StateTx::DATA_0_SCL_CLR:
+            gpio_clr(scl_);
+            ++stateTx_;
+            break;
+        case StateTx::ACK_SDA_SCL_SET:
```

Comment:
we need to start by clear scl set sda.
then wait a cycle.
then next cycle we set scl and sample sda for the ACK bit.
then wait a cycle

then we do the TX/RX state machine which will start by clearing SCL and (on tx) setting SDA.
the RX state machine should sample before the next high to low transition, where it should validate that the clock is actually high.